In [11]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import datetime

In [27]:
wdf = pd.read_csv('weather_features.csv')

In [3]:
wdf.shape

(178396, 17)

In [4]:
wdf.columns

Index(['dt_iso', 'city_name', 'temp', 'temp_min', 'temp_max', 'pressure',
       'humidity', 'wind_speed', 'wind_deg', 'rain_1h', 'rain_3h', 'snow_3h',
       'clouds_all', 'weather_id', 'weather_main', 'weather_description',
       'weather_icon'],
      dtype='object')

In [5]:
wdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178396 entries, 0 to 178395
Data columns (total 17 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   dt_iso               178396 non-null  object 
 1   city_name            178396 non-null  object 
 2   temp                 178396 non-null  float64
 3   temp_min             178396 non-null  float64
 4   temp_max             178396 non-null  float64
 5   pressure             178396 non-null  int64  
 6   humidity             178396 non-null  int64  
 7   wind_speed           178396 non-null  int64  
 8   wind_deg             178396 non-null  int64  
 9   rain_1h              178396 non-null  float64
 10  rain_3h              178396 non-null  float64
 11  snow_3h              178396 non-null  float64
 12  clouds_all           178396 non-null  int64  
 13  weather_id           178396 non-null  int64  
 14  weather_main         178396 non-null  object 
 15  weather_descripti

In [6]:
wdf.describe().transpose()

count         mean          std     min          25%      50%  \
temp        178396.0   289.618605     8.026199  262.24   283.670000   289.15   
temp_min    178396.0   288.330442     7.955491  262.24   282.483602   288.15   
temp_max    178396.0   291.091267     8.612454  262.24   284.650000   290.15   
pressure    178396.0  1069.260740  5969.631893    0.00  1013.000000  1018.00   
humidity    178396.0    68.423457    21.902888    0.00    53.000000    72.00   
wind_speed  178396.0     2.470560     2.095910    0.00     1.000000     2.00   
wind_deg    178396.0   166.591190   116.611927    0.00    55.000000   177.00   
rain_1h     178396.0     0.075492     0.398847    0.00     0.000000     0.00   
rain_3h     178396.0     0.000380     0.007288    0.00     0.000000     0.00   
snow_3h     178396.0     0.004763     0.222604    0.00     0.000000     0.00   
clouds_all  178396.0    25.073292    30.774129    0.00     0.000000    20.00   
weather_id  178396.0   759.831902   108.733223  200.00   800.000000   800.00   

                    75%          max  
temp         295.150000      315.600  
temp_min     293.730125      315.150  
temp_max     297.150000      321.150  
pressure    1022.000000  1008371.000  
humidity      87.000000      100.000  
wind_speed     4.000000      133.000  
wind_deg     270.000000      360.000  
rain_1h        0.000000       12.000  
rain_3h        0.000000        2.315  
snow_3h        0.000000       21.500  
clouds_all    40.000000      100.000  
weather_id   801.000000      804.000

In [7]:
wdf.head().transpose()

0                          1  \
dt_iso               2015-01-01 00:00:00+01:00  2015-01-01 01:00:00+01:00   
city_name                             Valencia                   Valencia   
temp                                   270.475                    270.475   
temp_min                               270.475                    270.475   
temp_max                               270.475                    270.475   
pressure                                  1001                       1001   
humidity                                    77                         77   
wind_speed                                   1                          1   
wind_deg                                    62                         62   
rain_1h                                      0                          0   
rain_3h                                      0                          0   
snow_3h                                      0                          0   
clouds_all                                   0                          0   
weather_id                                 800                        800   
weather_main                             clear                      clear   
weather_description               sky is clear               sky is clear   
weather_icon                               01n                        01n   

                                             2                          3  \
dt_iso               2015-01-01 02:00:00+01:00  2015-01-01 03:00:00+01:00   
city_name                             Valencia                   Valencia   
temp                                   269.686                    269.686   
temp_min                               269.686                    269.686   
temp_max                               269.686                    269.686   
pressure                                  1002                       1002   
humidity                                    78                         78   
wind_speed                                   0                          0   
wind_deg                                    23                         23   
rain_1h                                      0                          0   
rain_3h                                      0                          0   
snow_3h                                      0                          0   
clouds_all                                   0                          0   
weather_id                                 800                        800   
weather_main                             clear                      clear   
weather_description               sky is clear               sky is clear   
weather_icon                               01n                        01n   

                                             4  
dt_iso               2015-01-01 04:00:00+01:00  
city_name                             Valencia  
temp                                   269.686  
temp_min                               269.686  
temp_max                               269.686  
pressure                                  1002  
humidity                                    78  
wind_speed                                   0  
wind_deg                                    23  
rain_1h                                      0  
rain_3h                                      0  
snow_3h                                      0  
clouds_all                                   0  
weather_id                                 800  
weather_main                             clear  
weather_description               sky is clear  
weather_icon                               01n

In [8]:
wdf.isnull().sum()

dt_iso                 0
city_name              0
temp                   0
temp_min               0
temp_max               0
pressure               0
humidity               0
wind_speed             0
wind_deg               0
rain_1h                0
rain_3h                0
snow_3h                0
clouds_all             0
weather_id             0
weather_main           0
weather_description    0
weather_icon           0
dtype: int64

In [76]:
wdf_std = wdf.std()
wdf_std

temp             8.024910
temp_min         7.948249
temp_max         8.613916
pressure      6021.768900
humidity        21.838097
wind_speed       2.095140
wind_deg       116.548788
rain_1h          0.385915
rain_3h          0.007348
snow_3h          0.224547
clouds_all      30.339522
dtype: float64

In [9]:
wdf['city_name'].value_counts()

Madrid        36267
Bilbao        35951
Seville       35557
 Barcelona    35476
Valencia      35145
Name: city_name, dtype: int64

Now these numbers are **not** what we wanted to see. We want to see 35064 hours for each city (over 4 years, $4 \times 365.25 \times 24 = 35064$ hours), as in the previous energy dataset.

In [10]:
valencia_repeats = wdf.loc[wdf['city_name'] == 'Valencia', 'dt_iso'].value_counts()
valencia_repeats = valencia_repeats[valencia_repeats != 1].index

In [11]:
vr_data = wdf[(wdf['dt_iso'].isin(valencia_repeats)) & (wdf['city_name'] == 'Valencia')]

In [12]:
vr_data.head(30)

dt_iso city_name     temp  temp_min  temp_max  \
124   2015-01-06 04:00:00+01:00  Valencia  273.142   273.142   273.142   
125   2015-01-06 04:00:00+01:00  Valencia  273.142   273.142   273.142   
3981  2015-06-15 21:00:00+02:00  Valencia  289.610   288.650   290.930   
3982  2015-06-15 21:00:00+02:00  Valencia  289.610   288.650   290.930   
3983  2015-06-15 22:00:00+02:00  Valencia  289.240   288.150   290.370   
3984  2015-06-15 22:00:00+02:00  Valencia  289.240   288.150   290.370   
4773  2015-07-18 19:00:00+02:00  Valencia  299.770   298.750   300.930   
4774  2015-07-18 19:00:00+02:00  Valencia  299.770   298.750   300.930   
5090  2015-07-31 23:00:00+02:00  Valencia  295.890   293.710   297.590   
5091  2015-07-31 23:00:00+02:00  Valencia  295.890   293.710   297.590   
5092  2015-08-01 00:00:00+02:00  Valencia  295.340   293.710   297.150   
5093  2015-08-01 00:00:00+02:00  Valencia  295.340   293.710   297.150   
5236  2015-08-06 23:00:00+02:00  Valencia  298.990   297.040   302.150   
5237  2015-08-06 23:00:00+02:00  Valencia  298.990   297.040   302.150   
5260  2015-08-07 22:00:00+02:00  Valencia  300.590   299.260   302.150   
5261  2015-08-07 22:00:00+02:00  Valencia  300.590   299.260   302.150   
5280  2015-08-08 17:00:00+02:00  Valencia  295.920   294.820   296.850   
5281  2015-08-08 17:00:00+02:00  Valencia  295.920   294.820   296.850   
5834  2015-08-31 18:00:00+02:00  Valencia  298.790   297.040   301.150   
5835  2015-08-31 18:00:00+02:00  Valencia  298.790   297.040   301.150   
6120  2015-09-12 15:00:00+02:00  Valencia  297.610   295.370   299.150   
6121  2015-09-12 15:00:00+02:00  Valencia  297.610   295.370   299.150   
6483  2015-09-27 17:00:00+02:00  Valencia  291.220   287.040   294.260   
6484  2015-09-27 17:00:00+02:00  Valencia  291.220   287.040   294.260   
6565  2015-10-01 02:00:00+02:00  Valencia  289.740   286.480   292.040   
6566  2015-10-01 02:00:00+02:00  Valencia  289.740   286.480   292.040   
7479  2015-11-08 02:00:00+01:00  Valencia  286.150   286.150   286.150   
7480  2015-11-08 02:00:00+01:00  Valencia  286.150   286.150   286.150   
7482  2015-11-08 04:00:00+01:00  Valencia  286.150   286.150   286.150   
7483  2015-11-08 04:00:00+01:00  Valencia  286.150   286.150   286.150   

      pressure  humidity  wind_speed  wind_deg  rain_1h  rain_3h  snow_3h  \
124        995        86           1       265      0.0      0.0      0.0   
125        995        86           1       265      0.0      0.0      0.0   
3981      1018       100           3       330      0.9      0.0      0.0   
3982      1018       100           3       330      0.9      0.0      0.0   
3983      1018       100           2       300      0.9      0.0      0.0   
3984      1018       100           2       300      0.9      0.0      0.0   
4773      1013        83           4       120      0.9      0.0      0.0   
4774      1013        83           4       120      0.9      0.0      0.0   
5090      1009        88           3        70      3.0      0.0      0.0   
5091      1009        88           3        70      3.0      0.0      0.0   
5092      1008        88           4        80      0.3      0.0      0.0   
5093      1008        88           4        80      0.3      0.0      0.0   
5236      1015        70           2       110      0.9      0.0      0.0   
5237      1015        70           2       110      0.9      0.0      0.0   
5260      1017        70           4       180      0.3      0.0      0.0   
5261      1017        70           4       180      0.3      0.0      0.0   
5280      1019        94           1         0      0.9      0.0      0.0   
5281      1019        94           1         0      0.9      0.0      0.0   
5834      1012        74           2       120      0.3      0.0      0.0   
5835      1012        74           2       120      0.3      0.0      0.0   
6120      1014        69           4        90      0.3      0.0      0.0   
6121      1014        69           4

The culprit of these repetitions are due to the last four categorical `weather_` variables. These need to separated as a different dataset from the continuous variables

Make a new dataset `cat_wdf`.

In [28]:
cat_variables = wdf.columns.difference(wdf.columns[2:13], sort = False)

In [29]:
cat_variables

Index(['dt_iso', 'city_name', 'weather_id', 'weather_main',
       'weather_description', 'weather_icon'],
      dtype='object')

In [30]:
wdf_cat = wdf[cat_variables]
wdf_cat

dt_iso city_name  weather_id weather_main  \
0       2015-01-01 00:00:00+01:00  Valencia         800        clear   
1       2015-01-01 01:00:00+01:00  Valencia         800        clear   
2       2015-01-01 02:00:00+01:00  Valencia         800        clear   
3       2015-01-01 03:00:00+01:00  Valencia         800        clear   
4       2015-01-01 04:00:00+01:00  Valencia         800        clear   
...                           ...       ...         ...          ...   
178391  2018-12-31 19:00:00+01:00   Seville         800        clear   
178392  2018-12-31 20:00:00+01:00   Seville         800        clear   
178393  2018-12-31 21:00:00+01:00   Seville         800        clear   
178394  2018-12-31 22:00:00+01:00   Seville         800        clear   
178395  2018-12-31 23:00:00+01:00   Seville         800        clear   

       weather_description weather_icon  
0             sky is clear          01n  
1             sky is clear          01n  
2             sky is clear          01n  
3             sky is clear          01n  
4             sky is clear          01n  
...                    ...          ...  
178391        sky is clear          01n  
178392        sky is clear          01n  
178393        sky is clear          01n  
178394        sky is clear          01n  
178395        sky is clear          01n  

[178396 rows x 6 columns]

In [40]:
wdf_cat.to_csv('weathercat_dataset.csv', index = False)

This can be cleaned later. For now we deal with the repeated observations in the main `wdf` dataset.

In [6]:
wdf.drop(wdf.columns[13:], axis = 1, inplace = True)

In [7]:
wdf

dt_iso city_name     temp  temp_min  temp_max  \
0       2015-01-01 00:00:00+01:00  Valencia  270.475   270.475   270.475   
1       2015-01-01 01:00:00+01:00  Valencia  270.475   270.475   270.475   
2       2015-01-01 02:00:00+01:00  Valencia  269.686   269.686   269.686   
3       2015-01-01 03:00:00+01:00  Valencia  269.686   269.686   269.686   
4       2015-01-01 04:00:00+01:00  Valencia  269.686   269.686   269.686   
...                           ...       ...      ...       ...       ...   
178391  2018-12-31 19:00:00+01:00   Seville  287.760   287.150   288.150   
178392  2018-12-31 20:00:00+01:00   Seville  285.760   285.150   286.150   
178393  2018-12-31 21:00:00+01:00   Seville  285.150   285.150   285.150   
178394  2018-12-31 22:00:00+01:00   Seville  284.150   284.150   284.150   
178395  2018-12-31 23:00:00+01:00   Seville  283.970   282.150   285.150   

        pressure  humidity  wind_speed  wind_deg  rain_1h  rain_3h  snow_3h  \
0           1001        77           1        62      0.0      0.0      0.0   
1           1001        77           1        62      0.0      0.0      0.0   
2           1002        78           0        23      0.0      0.0      0.0   
3           1002        78           0        23      0.0      0.0      0.0   
4           1002        78           0        23      0.0      0.0      0.0   
...          ...       ...         ...       ...      ...      ...      ...   
178391      1028        54           3        30      0.0      0.0      0.0   
178392      1029        62           3        30      0.0      0.0      0.0   
178393      1028        58           4        50      0.0      0.0      0.0   
178394      1029        57           4        60      0.0      0.0      0.0   
178395      1029        70           3        50      0.0      0.0      0.0   

        clouds_all  
0                0  
1                0  
2                0  
3                0  
4                0  
...            ...  
178391           0  
178392           0  
178393           0  
178394           0  
178395           0  

[178396 rows x 13 columns]

In [8]:
wdf.drop_duplicates(subset = ['dt_iso', 'city_name'], inplace = True)
wdf

dt_iso city_name     temp  temp_min  temp_max  \
0       2015-01-01 00:00:00+01:00  Valencia  270.475   270.475   270.475   
1       2015-01-01 01:00:00+01:00  Valencia  270.475   270.475   270.475   
2       2015-01-01 02:00:00+01:00  Valencia  269.686   269.686   269.686   
3       2015-01-01 03:00:00+01:00  Valencia  269.686   269.686   269.686   
4       2015-01-01 04:00:00+01:00  Valencia  269.686   269.686   269.686   
...                           ...       ...      ...       ...       ...   
178391  2018-12-31 19:00:00+01:00   Seville  287.760   287.150   288.150   
178392  2018-12-31 20:00:00+01:00   Seville  285.760   285.150   286.150   
178393  2018-12-31 21:00:00+01:00   Seville  285.150   285.150   285.150   
178394  2018-12-31 22:00:00+01:00   Seville  284.150   284.150   284.150   
178395  2018-12-31 23:00:00+01:00   Seville  283.970   282.150   285.150   

        pressure  humidity  wind_speed  wind_deg  rain_1h  rain_3h  snow_3h  \
0           1001        77           1        62      0.0      0.0      0.0   
1           1001        77           1        62      0.0      0.0      0.0   
2           1002        78           0        23      0.0      0.0      0.0   
3           1002        78           0        23      0.0      0.0      0.0   
4           1002        78           0        23      0.0      0.0      0.0   
...          ...       ...         ...       ...      ...      ...      ...   
178391      1028        54           3        30      0.0      0.0      0.0   
178392      1029        62           3        30      0.0      0.0      0.0   
178393      1028        58           4        50      0.0      0.0      0.0   
178394      1029        57           4        60      0.0      0.0      0.0   
178395      1029        70           3        50      0.0      0.0      0.0   

        clouds_all  
0                0  
1                0  
2                0  
3                0  
4                0  
...            ...  
178391           0  
178392           0  
178393           0  
178394           0  
178395           0  

[175320 rows x 13 columns]

There are 175320 rows ($35064 \times 5$ cities $= 175320$), so dropping duplicates was successful. We still need to convert time to a proper `datetime` format and deal with daylight saving time shenanigans, like the energy dataset.

In [10]:
wdf['dt_iso'].nunique()

35064

Similar to the energy dataset, a `date_range` can be set for the time column, but repeated 5 times for the 5 cities. However, a more elegant method can be used to do the same thing but without repeating 5 times.

In [12]:
one_hour = datetime.timedelta(hours = 1)

In [14]:
wdf.rename(columns = {'dt_iso': 'time'}, inplace = True)

In [15]:
wdf['time'] = pd.to_datetime(wdf['time'], infer_datetime_format = True, utc = True)

In [16]:
wdf.head()

time city_name     temp  temp_min  temp_max  pressure  \
0 2014-12-31 23:00:00+00:00  Valencia  270.475   270.475   270.475      1001   
1 2015-01-01 00:00:00+00:00  Valencia  270.475   270.475   270.475      1001   
2 2015-01-01 01:00:00+00:00  Valencia  269.686   269.686   269.686      1002   
3 2015-01-01 02:00:00+00:00  Valencia  269.686   269.686   269.686      1002   
4 2015-01-01 03:00:00+00:00  Valencia  269.686   269.686   269.686      1002   

   humidity  wind_speed  wind_deg  rain_1h  rain_3h  snow_3h  clouds_all  
0        77           1        62      0.0      0.0      0.0           0  
1        77           1        62      0.0      0.0      0.0           0  
2        78           0        23      0.0      0.0      0.0           0  
3        78           0        23      0.0      0.0      0.0           0  
4        78           0        23      0.0      0.0      0.0           0

First, a time difference of one hour is prepared. Second, the time column is renamed to match that of the energy dataset. Third, the timezone is converted to UTC so that all timestamps will have a common `+00:00` instead of two different `+01:00` and `+02:00`. Finally, one hour is added to all timestamps and the redundant `+00:00` UTC timezone information is stripped away.

In [17]:
wdf['time'] = wdf['time'] + one_hour
wdf['time'] = [str(x) for x in wdf['time']]
wdf['time'] = [x[:16] for x in wdf['time']]
wdf['time'] = pd.to_datetime(wdf['time'], infer_datetime_format = True)

In [19]:
wdf

time city_name     temp  temp_min  temp_max  pressure  \
0      2015-01-01 00:00:00  Valencia  270.475   270.475   270.475      1001   
1      2015-01-01 01:00:00  Valencia  270.475   270.475   270.475      1001   
2      2015-01-01 02:00:00  Valencia  269.686   269.686   269.686      1002   
3      2015-01-01 03:00:00  Valencia  269.686   269.686   269.686      1002   
4      2015-01-01 04:00:00  Valencia  269.686   269.686   269.686      1002   
...                    ...       ...      ...       ...       ...       ...   
178391 2018-12-31 19:00:00   Seville  287.760   287.150   288.150      1028   
178392 2018-12-31 20:00:00   Seville  285.760   285.150   286.150      1029   
178393 2018-12-31 21:00:00   Seville  285.150   285.150   285.150      1028   
178394 2018-12-31 22:00:00   Seville  284.150   284.150   284.150      1029   
178395 2018-12-31 23:00:00   Seville  283.970   282.150   285.150      1029   

        humidity  wind_speed  wind_deg  rain_1h  rain_3h  snow_3h  clouds_all  
0             77           1        62      0.0      0.0      0.0           0  
1             77           1        62      0.0      0.0      0.0           0  
2             78           0        23      0.0      0.0      0.0           0  
3             78           0        23      0.0      0.0      0.0           0  
4             78           0        23      0.0      0.0      0.0           0  
...          ...         ...       ...      ...      ...      ...         ...  
178391        54           3        30      0.0      0.0      0.0           0  
178392        62           3        30      0.0      0.0      0.0           0  
178393        58           4        50      0.0      0.0      0.0           0  
178394        57           4        60      0.0      0.0      0.0           0  
178395        70           3        50      0.0      0.0      0.0           0  

[175320 rows x 13 columns]

In [20]:
wdf['time'].nunique()

35064

In [23]:
wdf['city_name'].value_counts()

Madrid        35064
Seville       35064
 Barcelona    35064
Bilbao        35064
Valencia      35064
Name: city_name, dtype: int64

This ensures that time starts from `2015-01-01 00:00` up to `2018-12-31 23:00` for all 5 cities, while also reflecting the original Europe/Madrid timezone instead of being in the UTC.

Lastly, temperatures are recorded in the Kelvin scale for some reason, so these were changed to Celsius.

In [24]:
K = 273.15

In [ ]:
wdf['temp'] = wdf['temp'] - K
wdf['temp_min'] = wdf['temp_min'] - K
wdf['temp_max'] = wdf['temp_max'] - K

In [20]:
wdf.to_csv('clean_weather_dataset.csv', index = False)

Do the same for the categorical weather dataset.

In [31]:
wdf_cat

dt_iso city_name  weather_id weather_main  \
0       2015-01-01 00:00:00+01:00  Valencia         800        clear   
1       2015-01-01 01:00:00+01:00  Valencia         800        clear   
2       2015-01-01 02:00:00+01:00  Valencia         800        clear   
3       2015-01-01 03:00:00+01:00  Valencia         800        clear   
4       2015-01-01 04:00:00+01:00  Valencia         800        clear   
...                           ...       ...         ...          ...   
178391  2018-12-31 19:00:00+01:00   Seville         800        clear   
178392  2018-12-31 20:00:00+01:00   Seville         800        clear   
178393  2018-12-31 21:00:00+01:00   Seville         800        clear   
178394  2018-12-31 22:00:00+01:00   Seville         800        clear   
178395  2018-12-31 23:00:00+01:00   Seville         800        clear   

       weather_description weather_icon  
0             sky is clear          01n  
1             sky is clear          01n  
2             sky is clear          01n  
3             sky is clear          01n  
4             sky is clear          01n  
...                    ...          ...  
178391        sky is clear          01n  
178392        sky is clear          01n  
178393        sky is clear          01n  
178394        sky is clear          01n  
178395        sky is clear          01n  

[178396 rows x 6 columns]

In [34]:
wdf_cat.rename(columns = {'dt_iso': 'time'}, inplace = True)
wdf_cat['time'] = pd.to_datetime(wdf_cat['time'], utc = True, infer_datetime_format = True)
wdf_cat['time'] = wdf_cat['time'] + one_hour
wdf_cat['time'] = [str(x) for x in wdf_cat['time']]
wdf_cat['time'] = [x[:16] for x in wdf_cat['time']]
wdf_cat['time'] = pd.to_datetime(wdf_cat['time'], infer_datetime_format = True)

In [35]:
wdf_cat

time city_name  weather_id weather_main  \
0      2015-01-01 00:00:00  Valencia         800        clear   
1      2015-01-01 01:00:00  Valencia         800        clear   
2      2015-01-01 02:00:00  Valencia         800        clear   
3      2015-01-01 03:00:00  Valencia         800        clear   
4      2015-01-01 04:00:00  Valencia         800        clear   
...                    ...       ...         ...          ...   
178391 2018-12-31 19:00:00   Seville         800        clear   
178392 2018-12-31 20:00:00   Seville         800        clear   
178393 2018-12-31 21:00:00   Seville         800        clear   
178394 2018-12-31 22:00:00   Seville         800        clear   
178395 2018-12-31 23:00:00   Seville         800        clear   

       weather_description weather_icon  
0             sky is clear          01n  
1             sky is clear          01n  
2             sky is clear          01n  
3             sky is clear          01n  
4             sky is clear          01n  
...                    ...          ...  
178391        sky is clear          01n  
178392        sky is clear          01n  
178393        sky is clear          01n  
178394        sky is clear          01n  
178395        sky is clear          01n  

[178396 rows x 6 columns]

There would be no use to dropping duplicates timestamps since they occur due to more than one weather condition happenning in the same hour. For example, `weather_main` can be `rain` and `thunderstorm` in the same hour.

In [ ]:
wdf_cat.to_csv('clean_weathercat_dataset.csv', index = False)